Add support for parallel NVIDIA's GPU-accelerated direct sparse solver cuDSS solver[cudss-dev]

yuyangdai opened the pull request:
In this work, we add a class `CuDSSSolver` based on a GPU-accelerated direct sparse solver library for solving large sparse linear systems. This PR includes:

1. A new class `CuDSSSolver` wraps NVIDIA's GPU-accelerated direct sparse solver library [`cuDSS`](https://developer.nvidia.com/cudss) (≥ 0.6.0, CUDA 12.x+).
2. Add a variant of ex1p in the `./example/cudss` for the `CuDSSSolver` for testing the `CuDSSSolver`.
3. Modify the `CMakeLists` and `makefile` for `CuDSSSolver`:
    - MFEM_USER_CUDSS (default: NO)
    - CUDSS_DIR （default: @MFEM_DIR@/../cudss).
4. Before testing the cudss solver, please make sure to set the communication-layer. 
```bash
export CUDSS_COMM_LIB=${CUDSS_DIR}/lib/libcudss_commlayer_<backend>.so
```
<!--GHEX{"author":"yuyangdai","assignment":"2025-12-02T16:27:19","editor":"tzanio","id":5124,"reviewers":["dylan-copeland","jandrej","camierjs","helloworld922"],"merge":"2026-06-05T23:04:21","approval":"2026-06-05T23:02:34"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#5124](https://github.com/mfem/mfem/pull/5124) | @yuyangdai | @tzanio | @dylan-copeland + @jandrej + @camierjs + @helloworld922 | 12/2/25 | 6/5/26 | 6/5/26 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

Commit:
Update cmake files for cuDSS

yuyangdai committed:
Add cudss solver

yuyangdai committed:
Update the makefiles for CuDSSSolver

yuyangdai committed:
Update the doc files for CuDSSSolver

yuyangdai committed:
Using the `make style` to format the class CuDSSSolver and example/cudss/ex1p codes

yuyangdai committed:
Ignore the solution files of cudss sample run

yuyangdai committed:
Merge branch 'master' into cudss-dev

yuyangdai committed:
Merge branch 'master' into cudss-dev

dylan-copeland commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,181 @@
+{
+public:
+   /// Specify the type of matrix we are applying the solver to
+   enum MatType
```
Can this be `enum class`?

dylan-copeland commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,181 @@
+   };
+
+   /// Specify the view type of matrix we are applying the solver to
+   enum MatViewType
```
Can this be `enum class`?

yuyangdai replied:
We have used `enum class` in the latest commit [d56493c](https://github.com/mfem/mfem/pull/5124/commits/d56493c2c52d1b2d0699ddfeda1d1751e57709f9).

dylan-copeland commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,181 @@
+   int nnz = 0; // the number of non zeros
+   int row_start = 0;  // the first row index in CSR matrix operator
+   int row_end = 0;   // the end row index in CSR matrix operator
+   int n_loc = 0;    // the number of the rows in CSR matrix operator
```
This class is derived from `Solver`, which is derived from `Operator`, which has the member `height`. Is `n_loc` the same as `height`?

dylan-copeland commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,181 @@
+   /// Method for initialization
+   void Init(MPI_Comm comm_);
+
+   /// Method for the initialization of cudss Handle
```
`cudss` -> `cuDSS`?

yuyangdai replied:
We have updated the comments in the latest commit [d56493c](https://github.com/mfem/mfem/pull/5124/commits/d56493c2c52d1b2d0699ddfeda1d1751e57709f9).

yuyangdai committed:
Refactored enum classes MatType and MatViewType, updated variable names, and update some comments.

yuyangdai commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,181 @@
+   enum MatType
+   {
+      /// CUDSS_MTYPE_GENERAL: General matrix [default] $LDU$ factorization will be computed with optional local or global pivoting
+      UNSYMMETRIC = 0,
```
We have renamed this enumerator in the latest commit [d56493c](https://github.com/mfem/mfem/pull/5124/commits/d56493c2c52d1b2d0699ddfeda1d1751e57709f9).

yuyangdai committed:
Update the `n_global` variable name for global number of rows

pazner commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,181 @@
+
+private:
+   // MPI_Comm
+   MPI_Comm *mpi_comm = NULL;
```
I don't think this needs to be a pointer.

pazner commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,181 @@
+   // MPI_Comm
+   MPI_Comm *mpi_comm = NULL;
+   // MPI rank
+   int myid;
```
Unused variable?

yuyangdai replied:
The variable `myid` is unused and it has been removed in the commit [d93d36d](https://github.com/mfem/mfem/pull/5124/commits/d93d36d6d385782b6c86bb4129451bdd39bb9a78#:~:text=void%20CuDSSSolver%3A%3AInit(MPI_Comm%20comm_)).

pazner commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,181 @@
+
+   /// Method for configuring storage for distributed/centralized RHS and
+   /// solution
+   void InitRhsSol(int nrhs_) const;
```
In MFEM, usually written 'RHS'

```suggestion
   void InitRHSSol(int nrhs_) const;
```

Maybe a clearer name would be `SetNumRHS` or similar?

pazner commented on linalg/cudss.cpp:
```diff
@@ -0,0 +1,372 @@
+{
+namespace
+{
+void CheckCudssAPI(cudssStatus_t status, std::string msg)
```
```suggestion
void CheckCudssAPI(cudssStatus_t status, std::string_view msg)
```

(or `const std::string &msh`)

pazner replied:
Nice solution!

pazner commented on linalg/cudss.cpp:
```diff
@@ -0,0 +1,372 @@
+{
+   mpi_comm = (MPI_Comm *)malloc(sizeof(MPI_Comm));
+   mpi_comm[0] = comm_; // MPI_COMM_WORLD
+   MPI_Comm_rank(comm_, &myid);
```
`myid` does not appear to be needed; no reason to have `mpi_comm` as a pointer. Can remove `CuDSSSolver::Init` entirely, and just initialize `mpi_comm` in the constructor

pazner commented on linalg/cudss.cpp:
```diff
@@ -0,0 +1,372 @@
+
+void CuDSSSolver::SetOperator(const Operator &op)
+{
+   bool cuDSSObjectInitialized = (Ac.get() != nullptr);
```
```suggestion
   bool cuDSSObjectInitialized = (Ac != nullptr);
```

pazner commented on config/cmake/modules/FindCUDSS.cmake:
```diff
@@ -0,0 +1,6 @@
+mfem_find_package(CUDSS CUDSS CUDSS_DIR
+  "include" cudss.h "lib" cudss
+  "Paths to headers required by cuDSS."
+  "Libraries required by cuDSS.")
```
Newline

pazner reviewed (commented):
Thanks for this contribution @yuyangdai! I think this has the potential to be very useful... I left some review comments.

yuyangdai committed:
Update linalg/cudss.cpp and linalg/cudss.hpp

- Define a cuDSS error check macro, MFEM_CUDSS_CHECK(x);
- Rename the method from InitRhsSol to SetNumRHS;
- Utilize CuMemAlloc/ CuMemcpyDtoD instead of cudaMalloc/cudaMemcpy;
- Update MPI_Comm usage.

yuyangdai committed:
Update config/cmake/modules/FindCUDSS.cmake and examples/cudss/CMakeLists.txt
- Add a newline at end of file

yuyangdai committed:
Update examples/ex1p.cpp
- Add cudss solver into the ex1p.cpp

Commit:
Fix CuDSSSolver::ArrayMult() for single RHS

yuyangdai commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,181 @@
+
+#endif // MFEM_USE_MPI
+#endif // MFEM_USE_CUDSS
+#endif // MFEM_CUDSS
```
A newline at the end of the `linalg/cudss.hpp` has been added in commit [d93d36d](https://github.com/mfem/mfem/pull/5124/commits/d93d36d6d385782b6c86bb4129451bdd39bb9a78#:~:text=endif%20//%20MFEM_CUDSS-,%23endif%20//%20MFEM_CUDSS,-Footer).

yuyangdai commented on linalg/cudss.cpp:
```diff
@@ -0,0 +1,372 @@
+}
+} // namespace
+
+CuDSSSolver::CuDSSSolver(MPI_Comm comm_)
```
Thanks! We have applied the suggested change in commit [d93d36d](https://github.com/mfem/mfem/pull/5124/commits/d93d36d6d385782b6c86bb4129451bdd39bb9a78#:~:text=CuDSSSolver%3A%3ACuDSSSolver(MPI_Comm%20comm_)%20%3A%20mpi_comm(comm_)).

yuyangdai commented on linalg/cudss.cpp:
```diff
@@ -0,0 +1,372 @@
+   cudaError_t cuda_status = cudaSuccess;
+
+   // Destroy the system Matrix
+   if (Ac.get())
```
Thanks! We have applied the suggested change in commit [d93d36d](https://github.com/mfem/mfem/pull/5124/commits/d93d36d6d385782b6c86bb4129451bdd39bb9a78#:~:text=the%20system%20Matrix-,if%20(Ac.get()),-if%20(Ac))

yuyangdai commented on linalg/cudss.cpp:
```diff
@@ -0,0 +1,372 @@
+
+   if (csr_offsets_d != NULL)
+   {
+      cuda_status = cudaFree(csr_offsets_d);
```
The `CuMemFree` is used in the commit [d93d36d](https://github.com/mfem/mfem/pull/5124/commits/d93d36d6d385782b6c86bb4129451bdd39bb9a78#:~:text=CuMemFree(csr_offsets_d)%3B).

yuyangdai commented on examples/cudss/ex1p.cpp:
The option of cudss solver has been added to the ex1p in the commit [3f236b4](https://github.com/mfem/mfem/pull/5124/commits/3f236b406f4604622e8dcc77d68b5c0d58568362).

pazner replied:
Thanks! If this is sufficient, then the examples/cudss subdirectory can be deleted, and the associated changes to the build system can be reverted.

jandrej replied:
Agree

pazner commented on linalg/cudss.hpp:
```diff
@@ -17,8 +17,22 @@
 {
+// Function used by the macro MFEM_CUDSS_CHECK.
+void mfem_cudss_error(cudssStatus_t status, const char *expr, const char *func,
+                      const char *file, int line);
```
Can move the macro to the cpp file and delete the function declaration from the hpp file.

yuyangdai replied:
We have moved the macro to the cpp file; see the commit [eae1fa2](https://github.com/mfem/mfem/pull/5124/commits/eae1fa217b50f9f366e8826e0a3463d31566d744) for details.

pazner commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,192 @@
+      UPPER = 2,
+   };
+
+   CuDSSSolver(MPI_Comm comm);
```
Should probably delete the move constructor and move assignment operator (as well as the copy constructor and copy assignment operator, even though those are implicitly deleted because of unique_ptr).

It could be nice to implement move semantics for this class, but it is a little tricky because of the cudss data that need to be cleaned up, so just deleting the special member functions is fine I think.

yuyangdai replied:
The move constructor and move assignment operator have been added. Please see the commit  [eae1fa2](https://github.com/mfem/mfem/pull/5124/commits/eae1fa217b50f9f366e8826e0a3463d31566d744)  for details.

jandrej commented on INSTALL:
```diff
@@ -395,6 +395,13 @@ MFEM_USE_STRUMPACK = YES/NO
 
+MFEM_USE_CUDSS = YES/NO
+   Enable MFEM functionality based on the cuDSS library. Currently, this option
+   adds the class CuDSSSolver (a parallel sparse direct solver). When using cuDSS,
```
Spelling consistency

pazner replied:
It looks OK to me actually (the library is "cuDSS", the MFEM class is CuDSSSolver). Or is there something else?

yuyangdai replied:
Thanks to @pazner. The "cuDSS" is the library name, and "CuDSSSolver" is the MFEM class.

yuyangdai committed:
Add `CuDSSHandle` singleton and update `CuDSSSolver`

- Remove unused variable `myid`.
- Move `MFEM_CUDSS_CHECK` and `mfem_cudss_error` into linalg/cudss.cpp.
- Disable move copy constructor and move assignment.
- Introduce `CuDSSHandle` singleton to manage cuDSS handle lifetime.
- Rename `InitHandle` to `Init`

yuyangdai committed:
Update the examples/ex1p.cpp
- Add the global cudss handle before using the cudss solver.

yuyangdai commented on linalg/cudss.cpp:
```diff
@@ -0,0 +1,372 @@
+   // Checkout the cuDSS API
+   cudssStatus_t status = CUDSS_STATUS_SUCCESS;
+
+   status = cudssCreate(&handle);
```
We have introduced the CuDSSHandle singleton to own a global cudss handle's lifetime. Users only need to call `CuDSSHandle::Init()` once before creating any `CuDSSSolver`; the singleton finalizes the handle when the `Device` instance is destroyed. See the commit [eae1fa2](https://github.com/mfem/mfem/pull/5124/commits/eae1fa217b50f9f366e8826e0a3463d31566d744) for details.

yuyangdai committed:
Revert `makefile` and `example/CMakeLists.txt`; delete `examples/cudss`

yuyangdai committed:
Revert code docs; add constructor comment; update the `.gitignore`

yuyangdai committed:
Update linalg/cudss.hpp and linalg/cudss.cpp; Revert general/device.cpp

- Move ` CUDA_REAL_T` to the cpp file.
- Remove `CuDSSHandle` singleton and replace it with a static cudssHandle_t variable.
- Delete `CuDSSHandle::Init()` from ex1p

yuyangdai commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,224 @@
+#define CUDA_REAL_T CUDA_R_32F
+#else
+#define CUDA_REAL_T CUDA_R_64F
+#endif
```
These have been moved to the cpp file; see commit [a8eba59](https://github.com/mfem/mfem/pull/5124/changes/a8eba594d257c17bbe1d7f2d814dcc4e7fa9fc19) for details.

kvoronin commented on CMakeLists.txt:
```diff
@@ -433,6 +433,15 @@ if (MFEM_USE_STRUMPACK)
+  if (MFEM_USE_CUDA AND MFEM_USE_MPI)
+    find_package(CUDSS REQUIRED)
+  else()
+    message(FATAL_ERROR " *** cuDSS requires that CUDA and MPI be enabled.")
```
FWIW, cuDSS does not need MPI necessarily, it can work just fine with a single GPU.
Also, FYI, starting with 0.7.0, we have MG mode (different from MGMN) which works for single node multiple GPU, without the need for MPI.

yuyangdai committed:
Make MPI optional and add OpenMP support in CuDSSSolver
- Make the MPI optional: the CuDSSSolver now does not need MPI necessary
- Add the OpenMP supports by cudssSetThreadingLayer() API

yuyangdai committed:
Add cuDSS solver option in ex1.cpp

yuyangdai committed:
Update parameter names in documentation for clarity in cudss.hpp

yuyangdai committed:
Add conditional support for cuDSS solver in ex1.cpp

camierjs committed:
Merge branch 'master' into cudss-dev

helloworld922 commented on examples/ex1.cpp:
```diff
@@ -225,9 +232,25 @@ int main(int argc, char *argv[])
    {
 #ifndef MFEM_USE_SUITESPARSE
+#ifdef MFEM_USE_CUDSS
+      if (cudss_solver && strstr(device_config, "cuda"))
```
```suggestion
      if (cudss_solver && Device::Allows(Backend::CUDA_MASK))
```

helloworld922 reviewed (changes requested):
Thanks for the contribution, just a few small changes to make.

yuyangdai committed:
Update config/cmake/modules/FindCUDSS.cmake

Co-authored-by: Andrew Ho <ho37@llnl.gov>

yuyangdai committed:
Update examples/ex1.cpp

Co-authored-by: Andrew Ho <ho37@llnl.gov>

yuyangdai committed:
Update examples/ex1p.cpp

Co-authored-by: Andrew Ho <ho37@llnl.gov>

yuyangdai committed:
Update examples/ex1p.cpp

Co-authored-by: Andrew Ho <ho37@llnl.gov>

yuyangdai committed:
Update linalg/cudss.hpp

Co-authored-by: Andrew Ho <ho37@llnl.gov>

yuyangdai committed:
Add dependency check for CUDSS in MFEMConfig.cmake.in

tzanio committed:
Merge branch 'master' into cudss-dev

camierjs committed:
Merge branch 'master' into cudss-dev

yuyangdai committed:
Make SetMatrix methods private and remove Init method
- SetMatrix methods are called inside SetOperator
- Init method has been removed as it is unnecessary

camierjs committed:
Merge branch 'master' into cudss-dev

camierjs committed:
Cosmetic trailing spaces

camierjs committed:
Add Copyright headers

camierjs committed:
Merge branch 'master' into cudss-dev

pazner commented on linalg/cudss.cpp:
```diff
@@ -0,0 +1,479 @@
+#endif // MFEM_USE_MPI
+   else
+   {
+      mfem_error("Unsupported Operator Type \n");
```
```suggestion
      MFEM_ABORT("Unsupported Operator Type \n");
```

pazner commented on examples/ex1p.cpp:
```diff
@@ -248,34 +253,51 @@ int main(int argc, char *argv[])
    }
-   else
+   // 13a. Solve using a direct solver with cuDSS
+#ifdef MFEM_USE_CUDSS
```
The logic here is a little tricky, but I think if the user sets `cudss_solver = true`, but in a non-CuDSS build, then no solver will be called at all?

I would reverse the order of these checks, so the CuDSS check comes first, and then the other case is just in an else branch, so you don't need to have both an if for `!pa && (cudss_solver && Device::Allows(Backend::CUDA_MASK))` and its negation `pa || !(cudss_solver && Device::Allows(Backend::CUDA_MASK))`.

camierjs committed:
Merge branch 'master' into cudss-dev

tzanio commented:
@yuyangdai, can you take a look and address the reviewer's comments?

yuyangdai committed:
Update cuDSS configuration and library paths in CMake and makefiles; refactor cuDSS solver integration.
- Add `MFEM_CUDSS_COMM_LIB` for OpenMPI communication library
- Add 'MFEM_CUDSS_THREADING_LIB' for threading library
- Add SetMatrixCuDSS() to set matrix values for cudss
- Remove SetMatrixSortRow()
- Remove unused options and simplify conditions in ex1.cpp and ex1p.cpp.

yuyangdai committed:
Fix CUDSS library checks and improve status output in CMake and makefile

yuyangdai committed:
Refactor cuDSS library path resolution in defaults.mk for improved handling of missing libraries

yuyangdai committed:
Refactor CuDSSSolver initialization.

yuyangdai committed:
Update INSTALL documentation for CUDSS requirements and backend support

yuyangdai commented on linalg/cudss.cpp:
```diff
@@ -0,0 +1,479 @@
+   MFEM_CUDSS_CHECK(cudssConfigCreate(&solverConfig));
+   MFEM_CUDSS_CHECK(cudssDataCreate(handle, &solverData));
+   MFEM_CUDSS_CHECK(cudssDataSet(handle, solverData, CUDSS_DATA_COMM,
+                                 &mpi_comm, sizeof(MPI_Comm *)));
```
We refactored the constructors by extracting the duplicated code into a new function `InitCuDSS`. See more details in commit [10840ac](https://github.com/mfem/mfem/pull/5124/changes/10840ac6b474032b08b05ff0d8bcf9989f7e163f).

pazner commented on examples/ex1p.cpp:
```diff
@@ -248,33 +248,51 @@ int main(int argc, char *argv[])
-   Solver *prec = NULL;
-   if (pa)
+#ifdef MFEM_USE_CUDSS
+   if (!pa && (Device::Allows(Backend::CUDA_MASK)))
```
Sorry for the back and forth on this! I think retaining the `-cudss` flag in ex1p is good, because we should stick with hypre BoomerAMG as the default (as opposed to serial ex1, where we default to direct solver when it's available).

pazner reviewed (commented):
This is getting very close, thanks @yuyangdai! Just a couple minor comments...

yuyangdai committed:
Update config/defaults.mk

Co-authored-by: Will Pazner <11493037+pazner@users.noreply.github.com>

yuyangdai committed:
Update cuDSS solver integration:
- Use the full name of default communication library and threading library.
- Check the `CUDSS_COMM_LIB` and `CUDSS_THREADING_LIB` in environment first.
- Add the `cudss-solver` option in ex1p

yuyangdai committed:
Add conditional compilation for cudss_solver in ex1p

pazner reviewed (approved):
Thanks @yuyangdai for this contribution! And thank you for your patience with so many back-and-forths during the review process.

camierjs committed:
Merge branch 'master' into cudss-dev

Pennycook commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,224 @@
+{
+public:
+   /// Specify the type of matrix we are applying the solver to
+   enum class MatType
```
Sorry for catching this late, but should this be `enum MatType` instead, to match the definition style in classes like `MUMPSSolver`?
https://github.com/mfem/mfem/blob/dbbd425a225b2ad0c3e1dc0e2b3992899e27556a/linalg/mumps.hpp#L38-L50

If this is an `enum class` then users have to write `CuDSSSolver::MatType::NONSYMMETRIC` instead of `CuDSSSolver::NONSYMMETRIC`. It's not clear to me what was intended, here: the example code shows the explicit `MatType::` scoping, but the docstrings for `SetMatrixSymType` and `SetMatrixViewType` do not.

Pinging @pazner for comment, since I don't know if the type definition or docstrings should be changed.

pazner replied:
I don't think we're entirely consistent on this. Either one would be fine by me.

Pennycook replied:
I can't mark this as resolved, but this was fixed in de8aacddce2e97f982736d1bbb44e62697818bd5.

yuyangdai replied:
The `enum class` has been replaced with `enum`. See commit for details [de8aacd](https://github.com/mfem/mfem/pull/5124/commits/de8aacddce2e97f982736d1bbb44e62697818bd5).

Pennycook commented on linalg/cudss.hpp:
```diff
@@ -0,0 +1,224 @@
+   };
+
+   /// Specify the view type of matrix we are applying the solver to
+   enum class MatViewType
```
Same comment as https://github.com/mfem/mfem/pull/5124/changes#r3091603785.

If the definition of `MatType` is changed, the definition of `MatViewType` should be changed as well.

yuyangdai replied:
The `enum class` has been replaced with `enum`. See commit for details [de8aacd](https://github.com/mfem/mfem/pull/5124/commits/de8aacddce2e97f982736d1bbb44e62697818bd5).

tzanio commented:
@yuyangdai, can you please resolve the conflicts?

yuyangdai committed:
Replace enum class with enum

yuyangdai committed:
Merge branch 'master' into cudss-dev

camierjs committed:
Merge branch 'master' into cudss-dev

tzanio commented:
@jandrej, @dylan-copeland, @v-dobrev and @camierjs -- any objections to merging?

camierjs committed:
Merge branch 'master' into cudss-dev

camierjs commented on examples/ex1p.cpp:
```diff
@@ -248,34 +253,51 @@ int main(int argc, char *argv[])
+         CuDSSSolver::MatType::SYMMETRIC_POSITIVE_DEFINITE);
+      cudss_solver.SetMatrixViewType(CuDSSSolver::MatViewType::UPPER);
+      cudss_solver.SetOperator(*A);
+      cudss_solver.Mult(B, X);
```
Using as `CUDSS_COMM_LIB ` the output of `sh ./cudss_build_commlayer.sh openmpi` with a specific MPI implementation allows to run in parallel.

v-dobrev commented on INSTALL:
```diff
@@ -719,6 +724,13 @@ The specific libraries and their options are:
 
+- CUDSS (optional), used when MFEM_USE_CUDSS = YES. Note that CUDSS requires 
+  CUDA 12.x toolkit and the cuDSS libraries. The supported communication backend
+  is OpenMPI 4.x (default), and OpenMPI 4.x or a later version must be pre-build.
```
Can we add a link or a sentence about how to use MPI implementations other than the one supported by the CUDSS distribution?

yuyangdai replied:
The link [“Communication layer library in cuDSS”](https://docs.nvidia.com/cuda/cudss/advanced_features.html#communication-layer-library-in-cudss) provides guidance on developing custom MPI implementations. This link has been added to the `INSTALL` file. See commit [59cef5f](https://github.com/mfem/mfem/pull/5124/commits/59cef5f9e3692b52533253e2b3e5a5032ab03b5c) for more details.

camierjs committed:
Merge branch 'master' into cudss-dev

yuyangdai committed:
Add a link for communication layer library in cuDSS

tzanio commented:
Merged in `next` for testing...

tzanio committed:
Merge branch 'master' into cudss-dev

v-dobrev commented on makefile:
```diff
@@ -742,6 +743,9 @@ status info:
 	$(info MFEM_USE_STRUMPACK     = $(MFEM_USE_STRUMPACK))
+	$(info MFEM_USE_CUDSS         = $(MFEM_USE_CUDSS))
+	$(info MFEM_CUDSS_COMM_LIB    = $(MFEM_CUDSS_COMM_LIB))
+	$(info MFEM_CUDSS_THREADING_LIB = $(MFEM_CUDSS_THREADING_LIB))
```
These two lines should not be printed here -- they are sub-options for cuDSS.

```suggestion
```

tzanio committed:
Update makefile

Co-authored-by: Veselin Dobrev <v-dobrev@users.noreply.github.com>

tzanio commented:
Re-merged in `next` for testing...

camierjs committed:
Merge branch 'master' into cudss-dev

tzanio reviewed (approved):
Thank you for this useful contribution @yuyangdai!

Can you please add an entry about it in the `CHANGELOG`?

tzanio committed:
Merge branch 'master' into cudss-dev

tzanio committed:
Merge branch 'master' into cudss-dev

yuyangdai committed:
Update CHANGELOG

camierjs committed:
Merge branch 'master' into cudss-dev

tzanio committed:
Merge branch 'master' into cudss-dev

tzanio commented:
@yuyangdai -- we are experimenting with using Copilot to help with MFEM reviews. Can you take a quick look at the suggestions above and let us know what you think (if they are not relevant, just feel free to resolve them).

yuyangdai commented on config/defaults.mk:
```diff
@@ -368,6 +369,19 @@ STRUMPACK_OPT = -I$(STRUMPACK_DIR)/include $(SCOTCH_OPT)
+CUDSS_OPT         = -I$(CUDSS_INCLUDE_DIR)
+CUDSS_LIB         = \
+ $(XLINKER)-rpath,$(CUDSS_LIBRARY_DIR) -L$(CUDSS_LIBRARY_DIR) -lcudss
+# The cuDSS communication and threading libraries. 
+MFEM_CUDSS_COMM_LIB = $(abspath $(wildcard $(or $(CUDSS_COMM_LIB),\
+   $(subst @MFEM_DIR@,$(MFEM_DIR), $(CUDSS_LIBRARY_DIR)/libcudss_commlayer_openmpi.so))))
+MFEM_CUDSS_THREADING_LIB = $(abspath $(wildcard $(or $(CUDSS_THREADING_LIB),\
+   $(subst @MFEM_DIR@,$(MFEM_DIR),$(CUDSS_LIBRARY_DIR)/libcudss_mtlayer_gomp.so))))
```
1. The program will first obtain the values of MFEM_CUDSS_COMM_LIB and MFEM_CUDSS_THREADING_LIB from environment variables.
2. The MFEM_CUDSS_COMM_LIB and MFEM_CUDSS_THREADING_LIB are used during the initialization of CuDSSSolver. If the directory does not exist, returning an empty value for the variable has the same effect as returning "NO".

I think these two commands are sufficient

yuyangdai commented on examples/ex1.cpp:
```diff
@@ -224,17 +224,29 @@ int main(int argc, char *argv[])
    // 11. Solve the linear system A X = B.
    if (!pa)
    {
+#ifdef MFEM_USE_CUDSS
+      if (Device::Allows(Backend::CUDA_MASK))
+      {
+         // Use cuDSS to solve the system.
+         CuDSSSolver cudss_solver;
+         cudss_solver.SetOperator(*A);
+         cudss_solver.Mult(B, X);
+      }
```
In this case, when the backend is CUDA and MFEM_USE_CUDSS is set to YES, cuDSS is the default solver.

helloworld922 commented on linalg/cudss.cpp:
```diff
@@ -0,0 +1,408 @@
+   bool cuDSSObjectInitialized = (Ac != nullptr);
+
+   const HypreParMatrix *A = dynamic_cast<const HypreParMatrix *>(&op);
+   hypre_ParCSRMatrix *parcsr_op = *A;
```
```suggestion
   hypre_ParCSRMatrix *parcsr_op = op;
   op.HypreRead();
   hypre_CSRMatrix *csr_op = hypre_MergeDiagAndOffd(parcsr_op);
```

helloworld922 commented on linalg/cudss.cpp:
```diff
@@ -0,0 +1,408 @@
+         // NOTE: For CuDSS solver to reuse the reordering (skipping analysis
+         // phase), it needs to access the I and J arrays of the **initial**
+         // matrix. Therefore, we need to copy and keep I and J in device memory.
+         CuMemAlloc(&csr_offsets_d, (n_loc + 1) * sizeof(int));
+         CuMemAlloc(&csr_columns_d, nnz * sizeof(int));
+
+         CuMemcpyDtoD(csr_offsets_d, csr_offsets, (n_loc + 1) * sizeof(int));
+         CuMemcpyDtoD(csr_columns_d, csr_columns, nnz * sizeof(int));
```
This is an internal function and only called with device pointers

helloworld922 commented:
@yuyangdai I looked through the copilot comments and it looks like one was relevant.

tzanio committed:
Merge branch 'master' into cudss-dev

yuyangdai committed:
Update linalg/cudss.cpp

Co-authored-by: Andrew Ho <ho37@llnl.gov>

yuyangdai committed:
Fix CMake script to correctly locate cuDSS library and allocate device memory for csr values when set the cudssMatrix.

yuyangdai commented on linalg/cudss.cpp:
```diff
@@ -0,0 +1,408 @@
+   }
+
+   cudaMalloc(&csr_values_d, nnz * sizeof(real_t));
+   CuMemcpyDtoD(csr_values_d, csr_values, nnz * sizeof(real_t));
```
When setting the cudssMatrix, allocate device memory for the CSR values. See commit [
b5e99d8](https://github.com/mfem/mfem/pull/5124/changes/b5e99d8d83c59bfe73fb93f8898b64f4d1984c95) for details.

tzanio commented:
Re-merged in `next` for testing...

tzanio commented:
In `INSTALL`, should `CUDSS_INCLUDE_LIB` be `CUDSS_LIB` to match the actual make variable?

yuyangdai committed:
Update INSTALL file to correct cuDSS library options

tzanio commented:
Re-merged in `next` for testing...

tzanio reviewed (approved):
@yuyangdai -- Thank you so much for this contribution and for your responsiveness during the review process! I think cuDSS will be really useful for many of our users, and I am looking forward to see what they do with it 🚀 

@helloworld922, @pazner and @camierjs -- Thank you so much for your careful review!

@Pennycook, @kvoronin, @v-dobrev, @dylan-copeland, @jandrej -- Thanks for jumping in to help!
